File: 백준/Gold/2638. 치즈/치즈.cc
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <algorithm>
using namespace std;

int chk[101][101];
int map[101][101];
int n,m,top,poll,cnt,ans;
int tx[4] = {-1,1,0,0};
int ty[4] = {0,0,-1,1};
struct queue{
    int x,y;
}q[10100];

void bfs(int x,int y) {
    top = poll = -1;
    q[++top].x = x;
    q[top].y = y;
    chk[x][y] = 1;
    while (top != poll) {
        int x1 = q[++poll].x;
        int y1 = q[poll].y;

        for (int i = 0; i < 4; ++i) {
            int nx = x1 + tx[i];
            int ny = y1 + ty[i];
            if(nx < 0 || nx >= n || ny < 0 || ny >= m){ continue;}
            if(!chk[nx][ny] && map[nx][ny] == 0){
                q[++top].x = nx;
                q[top].y = ny;
                chk[nx][ny]++;
            }else if(map[nx][ny]){
                chk[nx][ny]++;
            }

        }
     }
    


}

int main() {
    scanf("%d%d", &n, &m);

    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            scanf("%d",&map[i][j]);
        }
    }

    while(1){
        ans = 1;
        bfs(0,0);

        for (int i = 0; i < n; ++i) {
            for (int j = 0; j < m ; ++j) {
                if(chk[i][j]==1){
                    chk[i][j] = 0;
                }else if(chk[i][j] >= 2){
                    map[i][j] = 0;
                    chk[i][j] = 0;
                }
                if(map[i][j]){ans = 0;}
            }
        }


        cnt++;
        if(ans){printf("%d",cnt);return 0;}

    }

    return 0;
}

File: 백준/Gold/2636. 치즈/치즈.java
import java.io.*;
import java.util.*;

public class Main {
    static int [][]map = new int [101][101];
    static int [][]vitsit;
    static int []last = new int [200];
    static int [] tx = {-1,1,0,0};
    static int [] ty = {0,0,-1,1};
    static int n,m,ans,cnt;


    static public class Node{
        int x,y;
        public Node(int x, int y){
            this.x = x;
            this.y = y;
        }
    }


    static public void bfs(int x,int y){
        Queue<Node> q = new LinkedList<>();
        q.add(new Node(x,y));
        vitsit[x][y] = 1;

        while(!q.isEmpty()){
            Node node = q.poll();


            for (int i = 0; i < 4; i++) {
                int nx = node.x + tx[i];
                int ny = node.y + ty[i];
                if(nx < 0 || nx >= n || ny < 0 || ny >= m){continue;}
                if(map[nx][ny] == 0 && vitsit[nx][ny] == 0){
                    q.add(new Node(nx,ny));
                    vitsit[nx][ny] = 1;
                }
                if(map[nx][ny] == 1 && vitsit[nx][ny] == 0){
                    vitsit[nx][ny] = 2;
                    last[cnt] ++;
                }
            }

        }
    }

    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        st = new StringTokenizer(br.readLine());
        n = Integer.parseInt(st.nextToken());
        m = Integer.parseInt(st.nextToken());

        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            for (int j = 0; j < m; j++) {
                map[i][j] = Integer.parseInt(st.nextToken());
            }
        }

        while(true){
            vitsit = new int[101][101];
            ans = 0;
            bfs(0,0);
            for (int i = 0; i < n; i++) {
                for (int j = 0; j < m; j++) {
                    if(vitsit[i][j] == 2){
                        map[i][j] = 0;
                        ans = 1;
                    }
                }
            }
            if(ans == 0){
                break;
            }
            cnt++;
        }

        System.out.println(cnt+" "+last[cnt-1]);


    }
}